lib/livebook/hubs.ex

```diff
@@ -0,0 +1,130 @@
+defmodule Livebook.Hubs do
+  @moduledoc false
+
+  alias Livebook.Storage
+  alias Livebook.Hubs.{Fly, Metadata, Provider}
+
+  defmodule NotFoundError do
+    @moduledoc false
+
+    defexception [:id, plug_status: 404]
+
+    def message(%{id: id}) do
+      "could not find a hub matching #{inspect(id)}"
+    end
+  end
+
+  @namespace :hubs
+
+  @doc """
+  Gets a list of hubs from storage.
+  """
+  @spec fetch_hubs() :: list(Provider.t())
+  def fetch_hubs do
+    for fields <- Storage.current().all(@namespace) do
+      to_struct(fields)
+    end
+  end
+
+  @doc """
+  Gets a list of metadatas from storage.
+  """
+  @spec fetch_metadatas() :: list(Metadata.t())
+  def fetch_metadatas do
+    for hub <- fetch_hubs() do
+      Provider.normalize(hub)
+    end
+  end
+
+  @doc """
+  Gets one hub from storage.
+
+  Raises `NotFoundError` if the hub does not exist.
+  """
+  @spec fetch_hub!(String.t()) :: Provider.t()
+  def fetch_hub!(id) do
+    case Storage.current().fetch(@namespace, id) do
+      :error -> raise NotFoundError, id: id
+      {:ok, fields} -> to_struct(fields)
+    end
+  end
+
+  @doc """
+  Checks if hub already exists.
+  """
+  @spec hub_exists?(String.t()) :: boolean()
+  def hub_exists?(id) do
+    case Storage.current().fetch(@namespace, id) do
+      :error -> false
+      {:ok, _} -> true
+    end
+  end
+
+  @doc """
+  Saves a new hub to the configured ones.
+  """
+  @spec save_hub(Provider.t()) :: Provider.t()
+  def save_hub(struct) do
+    attributes = to_ets(struct)
+
+    with :ok <- Storage.current().insert(@namespace, struct.id, attributes),
+         :ok <- broadcast_hubs_change() do
```

Comment:
Should we also broadcast on deletion?

Reply:
The deletion is only for tests

Reply:
True, but it's effectively a part of the public internal API and I imagine we will also support deletion in the UI later :D

Reply:
Probably we will, so we will also start broadcasting. But since isn't the feature for this PR, we can keep things as is